util/movement-algs/src/grouping_heuristic/catch.rs

```diff
@@ -0,0 +1,43 @@
+use crate::grouping_heuristic::{GroupingHeuristic, GroupingOutcome};
+
+/// Heuristic combinator that converts all outcomes to failures if the inner heuristic fails.
+/// This relies on cloning the inner heuristic, so it is advised that this be used with types that can be cloned cheaply.
+pub struct CatchAllToFailure<T>(pub Box<dyn GroupingHeuristic<T>>)
+where
+	T: Clone;
+
+impl<T> CatchAllToFailure<T>
+where
+	T: Clone,
+{
+	pub fn new(heuristic: Box<dyn GroupingHeuristic<T>>) -> Self {
+		Self(heuristic)
+	}
+
+	pub fn boxed(heuristic: Box<dyn GroupingHeuristic<T>>) -> Box<Self> {
+		Box::new(Self(heuristic))
+	}
+}
+
+impl<T> GroupingHeuristic<T> for CatchAllToFailure<T>
+where
+	T: Clone,
+{
+	fn distribute(
+		&mut self,
+		distribution: Vec<GroupingOutcome<T>>,
+	) -> Result<Vec<GroupingOutcome<T>>, anyhow::Error> {
```

Comment:
I've got an idea to eliminate the need for cloning here: instead of `anyhow::Error`, make a `DistributeError` type that will give back the passed distribution:

```rust
pub struct DistributeError<T> {
    distribution: Vec<GroupingOutcome<T>>,
    // The actual error, delegated to for Display and Error impls
    cause: anyhow::Error,
}

impl<T> DistributeError<T> {
    /// Recovers the distribution that went into the `distribute` call.
    pub fn recover(self) -> Vec<GroupingOutcome<T>> {
        self.distribution
    }
}
```